url.go

```diff
@@ -251,28 +251,41 @@ func parseV1URL(url *gourl.URL, originalURL string) (*URL, error) {
 	return &r, nil
 }
 
-func (r *URL) path() string {
+func (u *URL) path() string {
 	var parts []string
-	if r.User != "" {
-		parts = append(parts, fmt.Sprintf("~%s", r.User))
+	if u.User != "" {
+		parts = append(parts, fmt.Sprintf("~%s", u.User))
 	}
-	if r.Series != "" {
-		parts = append(parts, r.Series)
+	if u.Series != "" {
+		parts = append(parts, u.Series)
 	}
-	if r.Revision >= 0 {
-		parts = append(parts, fmt.Sprintf("%s-%d", r.Name, r.Revision))
+	if u.Revision >= 0 {
+		parts = append(parts, fmt.Sprintf("%s-%d", u.Name, u.Revision))
 	} else {
-		parts = append(parts, r.Name)
+		parts = append(parts, u.Name)
```

Comment:
Curious as to how the store deals with URLs that have variable arguments... Does it pattern match on ~ for user and %s-%d for revision?

Reply:
As the charmstore uses the same code as this, or at least it did! Then it parsed it to extrapolate that same information that we had it. So in a sense, the URL would be isomorphic between juju and the charmstore. As we never pass the URL directly to the charmhub any longer, we are abusing the URL to just store the information we require, namely `<schema>:<name>-<revision>`. It has valid useful information in the URL and more importantly, it keeps Juju happy so that we don't have to re-write all the works that watch for the URL to change.

The longer term, this should be renamed to something other than URL, but that would be a mountain of work.

Reply:
Really? Guessing the meaning of args when parsing URLs is just a disaster waiting to happen... Any idea why these are not encoded as query string parameters in the first place?

Reply:
You see that would have been sane!

`<schema>://<username>@<name>?<revision>=&<series>=`